agora/Agora/Stake/Scripts.hs

```diff
@@ -345,39 +362,61 @@ mkStakeValidator impl sstSymbol pstClass gtClass =
-                    let allHaveSameOwner' =
-                          allHaveSameOwner
-                            #&& dF.owner
-                            #== firstStakeInputDatumF.owner
-                        allHaveSameOrOwnedByDelegatee' =
-                          let delegated =
-                                dF.delegatedTo #== firstStakeInputDatumF.delegatedTo
-                              ownedByDelegatee =
-                                pdata (pdjust # dF.owner)
-                                  #== firstStakeInputDatumF.delegatedTo
-                           in allHaveSameDelegatee
-                                #&& (delegated #|| ownedByDelegatee)
-                     in pcon $ PPair allHaveSameOwner' allHaveSameOrOwnedByDelegatee'
+            ( \stake helper -> unTermCont $ do
+                stakeF <- pletFieldsC @'["owner", "delegatedTo"] stake
+                helperF <- pmatchC helper
+
+                haveTheSameOwnerAsTheFirstStake <-
+                  pletC $
+                    stakeF.owner #== firstStakeInputDatumF.owner
+
+                let ownerOfTheFirstStakeIsTheDelegatee =
+                      haveTheSameOwnerAsTheFirstStake
+                        #|| stakeF.delegatedTo
+                        #== pdata (pdjust # firstStakeInputDatumF.owner)
+
+                    delegateeOfTheFirstStakeIsTheDelegatee =
+                      pdata (pdjust # stakeF.owner)
+                        #== firstStakeInputDatumF.delegatedTo
+                        #|| stakeF.delegatedTo
+                        #== firstStakeInputDatumF.delegatedTo
+
+                    helper' =
+                      pcon $
+                        PAuthCheckHelper
+                          ( helperF.allHaveTheSameOwner
+                              #&& haveTheSameOwnerAsTheFirstStake
+                          )
+                          ( helperF.allOwnByOrDelegatedToTheFirstOwner
+                              #&& ownerOfTheFirstStakeIsTheDelegatee
+                          )
+                          ( helperF.allOwnByOrDelegatedToTheFirstDelegatee
+                              #&& delegateeOfTheFirstStakeIsTheDelegatee
+                          )
+
+                pure helper'
+            )
+          # pcon
+            ( PAuthCheckHelper
+                (pconstant True)
+                (pconstant True)
+                (pconstant True)
             )
-          # pcon (PPair (pconstant True) (pconstant True))
           # restOfStakeInputDatums
 
     let ownerSignsTransaction =
-          allHaveSameOwner
+          authCheckHelperF.allHaveTheSameOwner
             #&& authorizedBy
             # firstStakeInputDatumF.owner
 
         delegateSignsTransaction =
-          allHaveSameOrOwnedByDelegatee
+          ( authCheckHelperF.allOwnByOrDelegatedToTheFirstOwner
+              #|| authCheckHelperF.allOwnByOrDelegatedToTheFirstDelegatee
+          )
```

Comment:
@chfanghr these two are two different code branches which need to check different signatures on L#423